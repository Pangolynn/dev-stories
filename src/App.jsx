// import * as React from "react";
import Button from "./UI/Button";

import React, { useEffect, useRef, useState } from "react";
import RadioButton from "./UI/RadioButton";
import Checkbox from "./UI/Checkbox";

const initStories = [
  {
    title: "React",
    url: "someUrl",
    author: "author1",
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: "Redux",
    url: "url2",
    author: "author2",
    num_comments: 2,
    points: 5,
    objectID: 1,
  },
];

const SET_STORIES = "SET_STORIES";
const REMOVE_STORY = "REMOVE_STORY";

const getAsyncStories = () =>
  new Promise((resolve) =>
    setTimeout(() => resolve({ data: { stories: initStories } }), 2000)
  );

// reducers come w/ a current state, and an action which comes with type/payload
const storiesReducer = (state, action) => {
  switch (action.type) {
    case SET_STORIES:
      // if action matches, return a new state
      return action.payload;
    case REMOVE_STORY:
      return state.filter(
        (story) => action.payload.objectID !== story.objectID
      );
    default:
      throw new Error();
  }
};

const useStorageState = (key, initialState) => {
  const [value, setValue] = useState(localStorage.getItem(key) || initialState);
  // set the initial state by accessing local Storage or initial state which is passed in

  useEffect(() => {
    // everytime the searchterm is updated (user types in search input) useffect
    // will run, and update the localStorage to match React's internal state.
    localStorage.setItem(key, value);
  }, [key, value]);

  // return an array to match the same way you could use useState hook
  return [value, setValue];
};

const App = () => {
  const [searchTerm, setSearchTerm] = useStorageState("search", "React");
  const [radioValue, setRadioValue] = useState("false");
  // const [stories, setStories] = useState([]);
  const [stories, dispatchStories] = React.useReducer(storiesReducer, []);
  const [isLoading, setLoading] = useState(false);
  const [isError, setError] = useState(false);

  useEffect(() => {
    setLoading(true);
    getAsyncStories()
      .then((result) => {
        dispatchStories({
          type: SET_STORIES,
          payload: result.data.stories,
        });
        setLoading(false);
      })
      .catch(() => setError(true));
  }, []);

  const onRadioChange = (e) => {
    setRadioValue((val) => {
      if (!val) {
        e.target.checked = false;
      }
      return !val;
    });
  };

  const handleSearch = (event) => {
    setSearchTerm(event.target.value);
  };

  const filteredStories = stories.filter((x) =>
    x.title.toLowerCase().includes(searchTerm.toLowerCase())
  );

  const handleRemoveStory = (item) => {
    dispatchStories({
      type: REMOVE_STORY,
      payload: item,
    });
  };

  return (
    <div>
      <h1>Dev Stories</h1>
      <InputWithLabel
        id="search"
        value={searchTerm}
        onInputChange={handleSearch}
        type="text"
        isFocused
      >
        Search:
      </InputWithLabel>
      <hr />

      {isError && <p>Something went wrong.</p>}
      {isLoading ? (
        <p>Loading</p>
      ) : (
        <List list={filteredStories} onRemoveItem={handleRemoveStory} />
      )}
      <Button classes="one two">Click Me</Button>
      <RadioButton
        value={radioValue}
        label="Test radio"
        onClick={onRadioChange}
      />
      <Checkbox name="testCheck" label="Test Check" />
    </div>
  );
};

const List = ({ list, onRemoveItem }) => (
  <ul>
    {list !== "" &&
      list.map((item) => (
        <>
          <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
        </>
      ))}
  </ul>
);

const Item = ({ item, onRemoveItem }) => {
  return (
    <li>
      {item.title}
      <span>
        <a href={item.url}>{item.title}</a>
      </span>
      <span>{item.author}</span>
      <span>{item.num_comments}</span>
      <span>{item.points}</span>
      <span>
        <Button type="button" onClick={() => onRemoveItem(item)}>
          Remove
        </Button>
      </span>
    </li>
  );
};

const InputWithLabel = ({
  id,
  children,
  value,
  onInputChange,
  type = "text",
  isFocused,
}) => {
  const inputRef = useRef();

  useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused, inputRef]);

  return (
    <div>
      <label htmlFor={id}>{children} </label>
      &nbsp;
      <input
        ref={inputRef}
        onChange={onInputChange}
        type={type}
        id={id}
        value={value}
      />
    </div>
  );
};

export default App;
