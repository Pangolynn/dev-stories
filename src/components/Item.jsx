import Button from "../UI/Button";
// import "./Item.css";

const Item = ({ item, onRemoveItem }) => {
  return (
    <li className="item">
      <span style={{ width: "40%" }}>
        <a href={item.url}>{item.title}</a>
      </span>
      <span style={{ width: "30%" }}>{item.author}</span>

      <span style={{ width: "10%" }}>{item.num_comments}</span>
      <span style={{ width: "10%" }}>{item.points}</span>
      <span style={{ width: "10%" }}>
        <Button
          type="button"
          onClick={() => onRemoveItem(item)}
          className="button button_small"
        >
          Remove
        </Button>
      </span>
    </li>
  );
};

export default Item;
