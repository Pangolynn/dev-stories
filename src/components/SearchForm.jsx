import InputWithLabel from "../UI/InputWithLabel";

const SearchForm = ({ searchTerm, onSearchInput, onSearchSubmit }) => (
  <form onSubmit={onSearchSubmit}>
    <InputWithLabel
      id="search"
      value={searchTerm}
      onInputChange={onSearchInput}
      type="text"
      isFocused
    >
      Search:
    </InputWithLabel>
    <button type="submit" disabled={!searchTerm}>
      Submit
    </button>
  </form>
);

export default SearchForm;
