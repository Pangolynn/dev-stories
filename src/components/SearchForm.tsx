import styles from "../App.module.css";

import InputWithLabel from "../UI/InputWithLabel";
import Button from "../UI/Button";

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit,
  className,
}) => (
  <form className={styles.searchForm} onSubmit={onSearchSubmit}>
    <InputWithLabel
      id="search"
      value={searchTerm}
      onInputChange={onSearchInput}
      type="text"
      isFocused
    >
      Search:
    </InputWithLabel>
    <Button className={className} type="submit" disabled={!searchTerm}>
      Submit
    </Button>
  </form>
);

export default SearchForm;
