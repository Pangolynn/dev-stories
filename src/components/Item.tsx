import Button from "../UI/Button";
import styles from "../App.module.css";
import { Story } from "../App";

type ItemProps = {
  item: Story;
  onRemoveItem: (item: Story) => void;
};

const Item: React.FC<ItemProps> = ({ item, onRemoveItem }) => {
  return (
    <li className={styles.item}>
      <span style={{ width: "40%" }}>
        <a href={item.url}>{item.title}</a>
      </span>
      <span style={{ width: "30%" }}>{item.author}</span>

      <span style={{ width: "10%" }}>{item.num_comments}</span>
      <span style={{ width: "10%" }}>{item.points}</span>
      <span style={{ width: "10%" }}>
        <Button
          type="button"
          onClick={() => onRemoveItem(item)}
          className={`${styles.button} ${styles.buttonSmall}`}
        >
          Dismiss
        </Button>
      </span>
    </li>
  );
};

export default Item;
