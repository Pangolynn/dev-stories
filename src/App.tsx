import axios from "axios";
import React, { useCallback, useEffect, useState } from "react";

// Styling
import styles from "./App.module.css";
import { ReactComponent as Github } from "./github.svg";

// UI Components
import Button from "./UI/Button";
// import RadioButton from "./UI/RadioButton";
import Checkbox from "./UI/Checkbox";

// Components
import List from "./components/List";
import SearchForm from "./components/SearchForm";

const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

/* storiesReducer action types */
const REMOVE_STORY = "REMOVE_STORY";
const STORIES_FETCH_INIT = "STORIES_FETCH_INIT";
const STORIES_FETCH_SUCCESS = "STORIES_FETCH_SUCCESS";
const STORIES_FETCH_FAILURE = "STORIES_FETCH_FAILURE";

type Story = {
  objectID: string;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};

type Stories = Story[];

type StoriesState = {
  data: Stories;
  isLoading: boolean;
  isError: boolean;
};

type StoriesFetchInitAction = {
  type: "STORIES_FETCH_INIT";
};

type StoriesFetchSuccessAction = {
  type: "STORIES_FETCH_SUCCESS";
  payload: Stories;
};

type StoriesFetchFailureAction = {
  type: "STORIES_FETCH_FAILURE";
};

type StoriesRemoveAction = {
  type: "REMOVE_STORY";
  payload: Story;
};

type StoriesAction =
  | StoriesFetchInitAction
  | StoriesFetchSuccessAction
  | StoriesFetchFailureAction
  | StoriesRemoveAction;

export type { Story, Stories };

const getLastSearches = (urls) => urls.slice(-5);

// reducers come w/ a current state, and an action which comes with type/payload
const storiesReducer = (state: StoriesState, action: StoriesAction) => {
  switch (action.type) {
    case STORIES_FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case STORIES_FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case STORIES_FETCH_FAILURE:
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case REMOVE_STORY:
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

const useStorageState = (
  key: string,
  initialState: string
): [string, (newValue: string) => void] => {
  const isMounted = React.useRef(false);

  const [value, setValue] = useState(localStorage.getItem(key) || initialState);
  // set the initial state by accessing local Storage or initial state which is passed in

  useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = true;
    } else {
      localStorage.setItem(key, value);
    }
  }, [key, value]);

  // return an array to match the same way you could use useState hook
  return [value, setValue];
};

const getSumComments = (stories) => {
  console.log("C");

  return stories.data.reduce((result, value) => result + value.num_comments, 0);
};

const App = () => {
  const [searchTerm, setSearchTerm] = useStorageState("search", "React");
  // const [radioValue, setRadioValue] = useState("false");

  const [stories, dispatchStories] = React.useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });

  const [urls, setUrls] = useState([`${API_ENDPOINT}${searchTerm}`]);

  // const onRadioChange = (e) => {
  //   setRadioValue((val) => {
  //     if (!val) {
  //       e.target.checked = false;
  //     }
  //     return !val;
  //   });
  // };

  const handleFetchStories = useCallback(async () => {
    dispatchStories({ type: STORIES_FETCH_INIT });

    try {
      const lastUrl = urls[urls.length - 1];
      const result = await axios.get(lastUrl);

      dispatchStories({
        type: STORIES_FETCH_SUCCESS,
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: STORIES_FETCH_FAILURE });
    }
  }, [urls]);

  useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    const url = `${API_ENDPOINT}${searchTerm}`;
    setUrls(urls.concat(url));

    event.preventDefault();
  };

  const handleRemoveStory = useCallback((item: Story) => {
    dispatchStories({
      type: REMOVE_STORY,
      payload: item,
    });
  }, []);

  const handleLastSearch = (url) => {
    // do something
  };

  const lastSearches = getLastSearches(urls);

  const sumComments = React.useMemo(() => getSumComments(stories), [stories]);

  return (
    <div className={styles.container}>
      <h1 className={styles.headlinePrimary}>
        <Github height="18px" width="18px" />
        Dev Stories with {sumComments}
      </h1>

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
        className="button-large"
      />

      {lastSearches.map((url) => (
        <button key={url} type="button" onClick={() => handleLastSearch(url)}>
          {url}{" "}
        </button>
      ))}

      {stories.isError && <p>Something went wrong.</p>}
      {stories.isLoading ? (
        <p>Loading</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </div>
  );
};

export default App;
export { storiesReducer };
