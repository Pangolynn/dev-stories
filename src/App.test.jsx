import { describe, it, expect, vi } from "vitest";
import App, { storiesReducer } from "./App";
import axios from "axios";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import Item from "./components/Item";
import SearchForm from "./components/SearchForm";

vi.mock("axios");

// Item,
// List,
// SearchForm,
// InputWithLabel

const storyOne = {
  title: "React",
  url: "https://reactjs.org/",
  author: "Jordan Walke",
  num_comments: 3,
  points: 4,
  objectID: 0,
};

const storyTwo = {
  title: "Redux",
  url: "https://redux.js.org/",
  author: "Dan Abramov, Andrw Clark",
  num_comments: 2,
  points: 5,
  objectID: 1,
};

const stories = [storyOne, storyTwo];

describe("storiesReducer", () => {
  it("removes a story from all stories", () => {
    const action = { type: "REMOVE_STORY", payload: storyOne };
    const state = { data: stories, isLoading: false, isError: false };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [storyTwo],
      isLoading: false,
      isError: false,
    };

    expect(newState).toStrictEqual(expectedState);
  });

  it("fails to fetch stories", () => {
    //given a state and an action, we expect the following new state
    const action = { type: "STORIES_FETCH_FAILURE", payload: [] };
    const state = { data: undefined, isLoading: false, isError: true };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: undefined,
      isLoading: false,
      isError: true,
    };

    expect(newState).toStrictEqual(expectedState);
  });

  it("fetches new stories", () => {
    const action = { type: "STORIES_FETCH_SUCCESS", payload: "ruby" };
    const state = { data: [], isLoading: false, isError: false };

    const newState = storiesReducer(state, action);

    // TODO: fix this, catching
    const getInit = async () => {
      try {
        const result = axios.get(
          "https://hn.algolia.com/api/v1/search?query=ruby"
        );

        const payload = result.data.hits;

        const expectedState = {
          data: payload,
          isLoading: true,
          isError: false,
        };

        expect(newState).toStrictEqual(expectedState);
      } catch (error) {
        console.log("error");
      }
    };
    getInit();
  });

  it("starts fetching stories", () => {
    const action = { type: "STORIES_FETCH_INIT" };
    const state = { data: [], isLoading: true, isError: false };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [],
      isLoading: true,
      isError: false,
    };

    expect(newState).toStrictEqual(expectedState);
  });
});

describe("App component", () => {
  it("removes an item when clicking the Dismiss button", () => {});
  it("requests some initial stories from an API", () => {});
});

describe("Item", () => {
  it("renders all properties", () => {
    render(<Item item={storyOne} />);

    expect(screen.getByText("Jordan Walke")).toBeInTheDocument();

    expect(screen.getByText("React")).toHaveAttribute(
      "href",
      "https://reactjs.org/"
    );
  });

  it("renders a clickable dismiss button", () => {
    render(<Item item={storyOne} />);

    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  it("clicking the dismiss button calls the callback handler", () => {
    const handleRemoveItem = vi.fn();

    render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);

    fireEvent.click(screen.getByRole("button"));

    expect(handleRemoveItem).toHaveBeenCalledTimes(1);
  });
});

describe("SearchForm", () => {
  const searchFormProps = {
    searchTerm: "react",
    onSearchInput: vi.fn(),
    onSearchSubmit: vi.fn(),
    className: "test",
  };

  // snapshot test
  it("renders snapshot", () => {
    const { container } = render(<SearchForm {...searchFormProps} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders the input field with its value", () => {
    render(<SearchForm {...searchFormProps} />);

    expect(screen.getByDisplayValue("react")).toBeInTheDocument();
  });

  it("renders the correct label", () => {
    render(<SearchForm {...searchFormProps} />);

    expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
  });

  it("calls onSearchInput on input field change", () => {
    render(<SearchForm {...searchFormProps} />);

    fireEvent.change(screen.getByDisplayValue("react"), {
      target: { value: "Redux" },
    });
    expect(searchFormProps.onSearchInput).toHaveBeenCalled(1);
  });

  it("calls onSearchSubmit on button submit click", () => {
    render(<SearchForm {...searchFormProps} />);

    fireEvent.submit(screen.getByRole("button"));

    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
  });
});

// TODO: InputWithLabel tests and List tests
describe("InputWithLabel", () => {
  it("renders an input field", () => {});

  it("renders a label", () => {});
});

describe("List", () => {
  it("renders a list of Items", () => {});
});

describe("App", () => {
  it("succeeds in fetching data", async () => {
    const promise = Promise.resolve({
      data: {
        hits: stories,
      },
    });

    axios.get.mockImplementationOnce(() => promise);

    render(<App />);

    expect(screen.queryByText(/Loading/)).toBeInTheDocument();

    await waitFor(async () => await promise);

    expect(screen.queryByText(/Loading/)).toBeNull();

    expect(screen.getByText("React")).toBeInTheDocument();

    expect(screen.getByText("Redux")).toBeInTheDocument();

    expect(screen.getAllByText("Dismiss").length).toBe(2);

    it("fails in fetching data", async () => {
      const promise = Promise.reject();

      axios.get.mockImplementationOnce(() => promise);

      render(<App />);

      expect(screen.getByText(/Loading/)).toBeInTheDocument();

      try {
        await waitFor(async () => await promise);
      } catch (error) {
        expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
      }
    });

    it("removes a story", async () => {
      const promise = Promise.resolve({
        data: {
          hits: stories,
        },
      });

      axios.get.mockImplementationOnce(() => promise);

      render(<App />);

      await waitFor(async () => await promise);

      expect(screen.getAllByText("Dismiss").length).toBe(2);
      expect(screen.getByText("Jordan Walke")).toBeInTheDocument();

      fireEvent.click(screen.getAllByText("Dismiss")[0]);

      expect(screen.getAllByText("Dismiss").length).toBe(1);
      expect(screen.queryByText("Jordan Walke")).toBeNull();
    });

    it("searches for specific stories", async () => {
      const reactPromise = Promise.resolve({
        data: {
          hits: stories,
        },
      });

      const anotherStory = {
        title: "JavaScript",
        url: "https://en.wikipedia.org/wiki/JavaScript",
        author: "Brendan Eich",
        num_comments: 15,
        points: 10,
        objectID: 3,
      };

      const javascriptPromise = Promise.resolve({
        data: {
          hits: [anotherStory],
        },
      });

      axios.get.mockImplementation((url) => {
        if (url.includes("React")) {
          return reactPromise;
        }

        if (url.includes("JavaScript")) {
          return javascriptPromise;
        }

        throw Error();
      });

      // initial render

      render(<App />);

      // First Data Fetching

      await waitFor(async () => await reactPromise);

      expect(screen.queryByDisplayValue("React")).toBeInTheDocument();

      expect(screen.queryByDisplayValue("JavaScript")).toBeNull();

      expect(screen.queryByText("Jordan Walke")).toBeInTheDocument();

      expect(
        screen.queryByText("Dan Abramov, Andre Clark")
      ).toBeInTheDocument();

      expect(screen.queryByText("Brendan Eich")).toBeNull();

      // user interaction -> search
      fireEvent.change(screen.queryByDisplayValue("React"), {
        target: {
          value: "JavaScript",
        },
      });
      expect(screen.queryByDisplayValue("React")).toBeNull();

      expect(screen.queryByDisplayValue("JavaScript")).toBeInTheDocument();

      fireEvent.submit(screen.queryByText("Submit"));

      // 2nd data fetching

      await waitFor(async () => await javascriptPromise);

      expect(screen.queryByText("Jordan Walke")).toBeNull();

      expect(screen.queryByText("Dan Abramov, Andrew Clark")).toBeNull();

      expect(screen.queryByText("Brendan Eich")).toBeInTheDocument();
    });
  });
});
