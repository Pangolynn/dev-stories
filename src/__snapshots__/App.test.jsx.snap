// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm > renders snapshot 1`] = `
<form
  class="_searchForm_6dc87e"
>
  <div>
    <label
      class="_label_6dc87e"
      for="search"
    >
      Search:
       
    </label>
    Â 
    <input
      class="_input_6dc87e"
      id="search"
      type="text"
      value="react"
    />
  </div>
  <button
    class="_button_6dc87e _test_6dc87e"
    type="submit"
  >
    Submit
  </button>
</form>
`;
